gemhardware/include/gem/hw/GEMHwDevice.h

```diff
@@ -581,10 +525,45 @@ namespace gem {
        * @brief Performs basic setup for the device
        * sets connection details (OBSOLETE)
        * sets logging level to Error
-       * Not inherited
+       * Not inherited, but calls a pure virtual function...
        **/
       void setup(std::string const& deviceName);
 
+      /**
+       * @brief Performs a check on the RPC response to verify whether there is an `error` key set
+       *        Every method that makes an RPC call *must* use this function to check the response
+       *
+       * @param caller should be the name of the function, only used in the error message
+       **/
+      void checkRPCResponse(std::string const& caller) const {
+        if (rsp.get_key_exists("error")) {
+          std::stringstream errmsg;
+          errmsg << rsp.get_string("error");
+          CMSGEMOS_ERROR(caller << ": " << errmsg.str());
+          XCEPT_RAISE(gem::hw::exception::RPCMethodError, errmsg.str());
+        }
+      }
+
+      /**
+       * @brief Extracts the device parameters from the device name
+       *
+       * @param deviceName should be of the format: gem-shelfXX-amcYY
+       *        optionally including -optohybridZZ
+       * @param index selects which of XX, YY, ZZ to return
+       **/
+      static uint8_t extractDeviceID(std::string const& deviceName, uint8_t const& index) {
+        std::vector<std::string> subs;
+        boost::split(subs, deviceName, boost::is_any_of("-"));
+        if (index < subs.size()) {
+          return stoull(subs[index].substr(subs[index].find_first_of("0123456789"),2),nullptr,10);
```

Comment:
What if the value is larger than 255? What if `find_first_of` can't find a digit?

Reply:
Actually, the limit is 99...

Good question, we'll have to see when someone grossly misuses this...